File: pet-services-api/internal/usecases/list_countries.go
package usecases

import (
	"context"
	"pet-services-api/internal/exceptions"
	"pet-services-api/internal/logging"
	"pet-services-api/internal/reference"
)

type ListCountriesOutput struct {
	Countries []reference.Country `json:"countries"`
}

type ListCountriesUseCase struct {
	service reference.Service
	logger  logging.LoggerInterface
}

func NewListCountriesUseCase(service reference.Service, logger logging.LoggerInterface) *ListCountriesUseCase {
	return &ListCountriesUseCase{service: service, logger: logger}
}

func (uc *ListCountriesUseCase) Execute(ctx context.Context) (*ListCountriesOutput, []exceptions.ProblemDetails) {
	uc.logger.LogInfo(ctx, "ListCountriesUseCase.Execute", "Listagem de países executada")
	return &ListCountriesOutput{Countries: uc.service.ListCountries()}, nil
}

File: pet-services-front/src/application/usecases/reference/list-countries.ts
import type { Country } from "@/domain";
import type { ReferenceGateway } from "@/application/ports";

export interface ListCountriesOutput {
  countries: Country[];
}

export class ListCountriesUseCase {
  constructor(private readonly gateway: ReferenceGateway) {}

  execute(): Promise<ListCountriesOutput> {
    return this.gateway.listCountries();
  }
}
